### PR TITLE
[Spanner] Refactoring session pool code to limit the lock scope when calculating batch size.

### PR DESCRIPTION
**Pull request by amanda-tarafa**
Proposal for addressing TODO from https://github.com/googleapis/google-cloud-dotnet/pull/3567#discussion_r337437131.

### CODE REVIEW

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs [hunk 1]
```diff
@@ -539,48 +544,21 @@ void RemoveStaleOrExpiredItemsFromStack(ConcurrentStack<PooledSession> stack)
...
             private void StartSessionCreationTasksIfNecessary()
             {
-                if (!Healthy || Shutdown)
-                {
-                    return;
-                }
-                // TODO: Refactor code so that _liveOrRequestedSessionCount and _inFlightSessionCreationCount are
-                // updated inside the lock without having all the calls to async methods.
-                lock (_lock)
+                // Let's start the loop on the background as well, since it acquieres the lock
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs]: It's that typo again :)

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs [hunk 2]
```diff
@@ -745,23 +735,65 @@ private async Task<IList<PooledSession>> CreatePooledSessionsAsync(int batchSize
...
+                            writeHealthChangedLog = wasHealthy != success;
                         }
+
+                        // We requested batchSize sessions to be created, but in reality actualCreatedSession were created, which are now live.
+                        // Note that actualCreatedSessions might be 0 either because the call failed or because no sessions were created.
+                        Interlocked.Add(ref _liveOrRequestedSessionCount, -batchSize + actualCreatedSessions);
+
+                        Interlocked.Add(ref _inFlightSessionCreationCount, -batchSize);
+                    }
+                    if(writeHealthChangedLog)
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs]: Formatting nit: space after if

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs [hunk 3]
```diff
@@ -745,23 +735,65 @@ private async Task<IList<PooledSession>> CreatePooledSessionsAsync(int batchSize
...
+                        // However, we don't want to end up with more than the maximum number of active sessions
+                        // in the options.
+
+                        // In reality, the current pool size should be 0 if pendingAcquisitionCount is non-zero, and
+                        // vice versa. However, we don't strictly enforce this, and the maths works out more simply if
+                        // we don't worry about it.
+
+                        // How many more sessions do we need to create in order to get the pool to the minimum size, after satisfying
+                        // pending callers, assuming that all the in-flight session requests succeed, and there are no more requests?
+                        int newSessionsToSatisfyPool = (_pendingAcquisitions.Count + Options.MinimumPooledSessions) - (poolSize + InFlightSessionCreationCount);
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs]: I do wonder whether we can use LiveOrRequestedSessionCount here, but I'm getting slightly lost in all the counters we have :(

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs (hunk 1 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs]: Done :(

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs (hunk 2 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs]: Done

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs (hunk 3 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.TargetedSessionPool.cs]: LiveOrRequestedSessionCount includes the _pendingAcquisitions ones if I remember correctly.
I do agree that we should be able to simplify this, and I was actually looking into doing it on this PR but decided against it to mitigate risks.
I'm adding a TODO on my TODO list to look into this further.

**Comment by amanda-tarafa** [comment]
Thanks! Addressed comments, will merge on green.

**Pull request merged**